src/coreclr/scripts/genEventPipe.py

```diff
@@ -284,28 +284,31 @@ def winTypeToFixedWidthType(t):
                 pack_list.append("#if BIGENDIAN")
                 pack_list.append("    const uint8_t *valuePtr = %s;" % paramName)
                 pack_list.append("    for (uint32_t i = 0; i < %s; ++i) {" % template.structs[paramName])
-                types = [winTypeToFixedWidthType(t) for t in template.structTypes[paramName]]
-                for t in set(types) - {"UTF8String", "UTF16String"}:
-                    pack_list.append("        %(type)s value_%(type)s;" % {'type': t})
-                if "UTF8String" in types or "UTF16String" in types:
-                    pack_list.append("        size_t value_len;")
-                for t in types:
-                    if t == "UTF8String":
-                        pack_list.append("        value_len = strlen((const char *)valuePtr);")
-                        pack_list.append("        success &= write_buffer_string_utf8_t((const ep_char8_t *)valuePtr, value_len, &buffer, &offset, &size, &fixedBuffer);")
-                        pack_list.append("        valuePtr += value_len + 1;")
-                    elif t == "UTF16String":
-                        pack_list.append("        value_len = strlen((const char *)valuePtr);")
-                        pack_list.append("        success &= write_buffer_string_utf8_to_utf16_t((const ep_char8_t *)valuePtr, value_len, &buffer, &offset, &size, &fixedBuffer);")
```

Comment:
This did a utf8 to utf16 conversion, and wrote the converted data into the buffer. It is now replaced by a call to write_buffer_string_utf16, but that assumes input is in utf16 since there is no utf8 -> utf16 conversion, something missing?

Reply:
My previous understanding was that the source string coming from Mono was UTF-8 encoded. Though, this does not seem to be true. For the BulkType event the source string is already in UTF-16 format (although at the moment only an empty string is emitted). Thus I'm only left with byte swaps for UTF-16 strings.

https://github.com/dotnet/runtime/blob/a34291586e5f82a98f6b5a2e9134cee96f7fc306/src/mono/mono/eventpipe/ep-rt-mono.c#L3187

Reply:
OK, probably a side effect of how the bulk type event is serialized, normally we pass strings as UTF8 down to FireETW layer where the conversion and serialization takes place directly into the buffer, but for the bulk type events it gets serialized above the FireETW layer, ep_rt_mono_fire_bulk_type_event on Mono and BulkTypeEventLogger::FireBulkTypeEven on CoreCLR. On Mono we represent unicode strings as UTF8 and then on serialize layer we do the conversion, UTF8->UTF16, but as said since the serialization layer for this event type is partly moved up on level causing issues. When we serialize structs, we normally use the the ep_write_buffer_x that would already do byte swapping when serializing into the buffer, so maybe it would be better to fix this when where we serialize the structs instead of after in the serialized byte stream as done below? Sounds like this is mainly an issue in how bulk type event is serialized and that it should use existing ep_write_buffer_x instead of its own versions.

Reply:
Thanks for the clarification. I will come up with an implementation which makes use of `write_buffer_x` in the FireETW layer for the bulk type event.

Reply:
What I think we should do is drop all the conversion done in the FireETW layer for structs that are part of a byte buffera and fix ep_rt_mono_fire_bulk_type_event to use our ep_write_buffer_x versions instead of its own buffer serializing functions unaware of endieness. The existing structs should already be written into the byte arrays using the ep_write_buffer_x so shouldn't need any additional work and could be serialized as is.